Source/WebCore/platform/ScrollSnapAnimatorState.cpp

```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+}
+
 bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const ScrollExtents& scrollExtents, float pageScale)
 {
+    // Check if we need to set the active target to a focused element
+    setFocusedElementForAxis(ScrollEventAxis::Vertical);
+    setFocusedElementForAxis(ScrollEventAxis::Horizontal);
+
     bool snapPointChanged = false;
-    // If we are already snapped in a particular axis, maintain that. Otherwise, snap to the nearest eligible snap point.
     auto activeHorizontalIndex = activeSnapIndexForAxis(ScrollEventAxis::Horizontal);
-    if (!activeHorizontalIndex || *activeHorizontalIndex >= snapOffsetsForAxis(ScrollEventAxis::Horizontal).size())
-        snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Horizontal, scrollOffset, scrollExtents, pageScale);
-
     auto activeVerticalIndex = activeSnapIndexForAxis(ScrollEventAxis::Vertical);
-    if (!activeVerticalIndex || *activeVerticalIndex >= snapOffsetsForAxis(ScrollEventAxis::Vertical).size())
+    auto activeHorizontalID = activeSnapIDForAxis(ScrollEventAxis::Horizontal);
+    auto activeVerticalID = activeSnapIDForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsVertical = snapOffsetsForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsHorizontal = snapOffsetsForAxis(ScrollEventAxis::Horizontal);
+    
+    // Check if we need to set the current indeces
+    if (!activeVerticalIndex || *activeVerticalIndex >= snapOffsetsForAxis(ScrollEventAxis::Vertical).size()) {
         snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Vertical, scrollOffset, scrollExtents, pageScale);
-
+        activeVerticalIndex = activeSnapIndexForAxis(ScrollEventAxis::Vertical);
+    }
+    if (!activeHorizontalIndex || *activeHorizontalIndex >= snapOffsetsForAxis(ScrollEventAxis::Horizontal).size()) {
+        snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Horizontal, scrollOffset, scrollExtents, pageScale);
+        activeHorizontalIndex = activeSnapIndexForAxis(ScrollEventAxis::Horizontal);
+    }
+    
+    // If we have an active target, see if we need to preserve it
+    if (activeHorizontalID && activeHorizontalIndex) {
+        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
+            preserveCurrentTargetForAxis(ScrollEventAxis::Horizontal);
+    } else if (activeVerticalID && activeVerticalIndex) {
+        if (*activeVerticalID != snapOffsetsVertical[*activeVerticalIndex].snapTargetId)
+            preserveCurrentTargetForAxis(ScrollEventAxis::Vertical);
+    }
+
+    // If we are snapped to multiple targets, save them
+    if ((!activeHorizontalID && activeHorizontalIndex) && (!activeVerticalID && activeVerticalIndex) && snapOffsetsForAxis(ScrollEventAxis::Horizontal)[*activeHorizontalIndex].snapTargetId != snapOffsetsForAxis(ScrollEventAxis::Vertical)[*activeVerticalIndex].snapTargetId) {
```

Comment:
This should probably join the if-else-if sequence above, just to make it clearer that this will not happen if we are preserving a previous snap point. I was a bit thrown off by this at first, which is why I mention it.

Reply:
At least to me it makes more sense to have them separate, as the first set of checks are for preserving the target for a particular axis, while the second check is for setting the targets if we don't have any. These seem logically separate to me. I updated the comment to make this difference a bit clearer.